File: src/846_hand_of_straights/hand_of_straights.cpp
#include <iostream>
#include <map>
#include <vector>

using namespace std;

class Solution {
   public:
    bool isNStraightHand(vector<int>& hand, int groupSize) {
        if (hand.size() % groupSize != 0) return false;
        map<int, int> count;
        for (int h : hand) count[h]++;
        while (!count.empty()) {
            int start = count.begin()->first;
            int cnt = count.begin()->second;
            for (int i = 0; i < groupSize; ++i) {
                int key = start + i;
                if (!count.count(key)) return false;
                if (count[key] < cnt) return false;
                count[key] -= cnt;
                if (count[key] == 0) count.erase(key);
            }
        }
        return true;
    }
};

int main() {
    Solution sol;
    vector<int> h1 = {1, 2, 3, 6, 2, 3, 4, 7, 8};
    cout << (sol.isNStraightHand(h1, 3) ? "True" : "False") << " | True" << endl;
    vector<int> h2 = {1, 2, 3, 4, 5};
    cout << (sol.isNStraightHand(h2, 4) ? "True" : "False") << " | False" << endl;
    vector<int> h3 = {1, 1, 2, 2, 3, 3};
    cout << (sol.isNStraightHand(h3, 3) ? "True" : "False") << " | True" << endl;
    vector<int> h4 = {1, 2, 3, 4};
    cout << (sol.isNStraightHand(h4, 2) ? "True" : "False") << " | True" << endl;
    vector<int> h5 = {1};
    cout << (sol.isNStraightHand(h5, 1) ? "True" : "False") << " | True" << endl;
    return 0;
}

File: src/846_hand_of_straights/HandOfStraights.java
import java.util.*;

class Solution {
    public boolean isNStraightHand(int[] hand, int groupSize) {
        if (hand.length % groupSize != 0)
            return false;
        TreeMap<Integer, Integer> count = new TreeMap<>();
        for (int h : hand)
            count.put(h, count.getOrDefault(h, 0) + 1);
        while (!count.isEmpty()) {
            int start = count.firstKey();
            int cnt = count.get(start);
            for (int i = 0; i < groupSize; ++i) {
                int key = start + i;
                if (!count.containsKey(key))
                    return false;
                int val = count.get(key);
                if (val < cnt)
                    return false;
                if (val == cnt)
                    count.remove(key);
                else
                    count.put(key, val - cnt);
            }
        }
        return true;
    }
}

public class HandOfStraights {
    public static void main(String[] args) {
        Solution sol = new Solution();
        System.out.println(sol.isNStraightHand(new int[] { 1, 2, 3, 6, 2, 3, 4, 7, 8 }, 3) + " | True");
        System.out.println(sol.isNStraightHand(new int[] { 1, 2, 3, 4, 5 }, 4) + " | False");
        System.out.println(sol.isNStraightHand(new int[] { 1, 1, 2, 2, 3, 3 }, 3) + " | True");
        System.out.println(sol.isNStraightHand(new int[] { 1, 2, 3, 4 }, 2) + " | True");
        System.out.println(sol.isNStraightHand(new int[] { 1 }, 1) + " | True");
    }
}
